### ISSUE TITLE
FIREBASE_CONFIG environment variable

### ISSUE THREAD

**Issue by arkum** [google-github-actions/deploy-cloud-functions#17]
### How do i set FIREBASE_CONFIG environment variable for my python function
I am not able to set the FIREBASE_CONFIG environment variable using the KEY1=VALUE syntax using env_var arguments of the google-github-actions/deploy-cloud-functions@main action. 
I get the below error

Env Vars must be in "KEY1=VALUE1,KEY2=VALUE2" format, received "databaseURL":"https://<projectid>.firebaseio.com"

**Comment by bharathkkb** [comment]
Hi @arkum 
Thanks for the report. Is the `env_vars: databaseURL=https://firebaseio.com` syntax throwing the error?
Could you show an example config?

**Comment by arkum** [comment]
Thanks @bharathkkb The expected format is as below

Key is `FIREBASE_CONFIG` and Value is `{"projectId":"project-id","databaseURL":"https://project-id.firebaseio.com","storageBucket":"project-id.appspot.com","locationId":"asia-south1"}
`
Where project-id will the gcp/firebase projectid

`
env_vars:FIREBASE_CONFIG={"projectId":"project-id","databaseURL":"https://project-id.firebaseio.com","storageBucket":"project-id.appspot.com","locationId":"asia-south1"}
`
i tried passing the value inside a single quote and double quote. But did not work

**Comment by bharathkkb** [comment]
@arkum that makes sense. What is happening is we split KEY=VALUE pairs based on `,` for env_vars and it is trying to parse by splitting the json configs which also contains `,`.  A short term fix might to break up into individual k,v pairs i.e `env_vars: projectId= project-id,databaseURL=https://project-id.firebaseio.com,storageBucket=project-id.appspot.com` and then in your application, create the `FIREBASE_CONFIG`. For instance,
```ts
const FIREBASE_CONFIG={};
FIREBASE_CONFIG.projectId=process.env.projectId;
FIREBASE_CONFIG.databaseURL=process.env.databaseURL;
...
```

Longer term I think we should have a better solution, perhaps we use a less common delimiter or have an escape char. @averikitsch I assume this is also relevant for Run, WDYT?

**Comment by arkum** [comment]
thank you @bharathkkb i  can see gcloud functions deploy has option for reading environment variables from a file. Wondering whether we can bring it to Function deploy action. Another thing i want to try is to use the setup gcloud action and try to run the gcloud function set-env-var command. I will keep you posted

**Comment by arkum** [comment]
Hi @bharathkkb deploying through gcloud works
`- name: Setup gcloud


    uses: google-github-actions/setup-gcloud@master
        with:
          project_id: project-id
          service_account_key: ${{ secrets.FIREBASE_FUNCTION_TEST_SERVICE_ACCOUNT }}
          export_default_credentials: true  
      - name: Deploy poincareplot function
        working-directory: ./Python/PythonFunction
        run: gcloud functions deploy functioname --runtime=python38 --entry-point=entrypoint --env-vars-file ../config-test.yml
`
In my config file i defined the environment variable as 

FIREBASE_CONFIG: '{"projectId":"project-id","databaseURL":"https://project-id.firebaseio.com","storageBucket":"project-id.appspot.com","locationId":"asia-south1"}'

**Comment by bharathkkb** [comment]
Thanks for the update @arkum 
env-vars-file seems like a reasonable add and would keep the configs cleaner

**Comment by averikitsch** [comment]
> Longer term I think we should have a better solution, perhaps we use a less common delimiter or have an escape char. @averikitsch I assume this is also relevant for Run, WDYT?

Passing in JSON for the FIREBASE_CONFIG seems like an anti-pattern. I would only use the firebase CLI to interact and create this config. It also looks like this is a [special env var](https://firebase.google.com/docs/functions/config-env#automatically_populated_environment_variables) functions uses and I wouldn't try to overwrite it. These values should be passed into the action or used with setup-gcloud.

A solution for both would be looking into [escaping quotes for the JSON value](https://stackoverflow.com/questions/59805337/google-cloud-run-and-json-in-env-variables) or being able to add a [delimiter](https://cloud.google.com/sdk/gcloud/reference/topic/escaping). Since env vars may contain sensitive information, I wouldn't  want to commit them to a file.

**Comment by bharathkkb** [comment]
> It also looks like this is a [special env var](https://firebase.google.com/docs/functions/config-env#automatically_populated_environment_variables) functions uses and I wouldn't try to overwrite it. These values should be passed into the action or used with setup-gcloud.

That makes sense, I did not know `FIREBASE_CONFIG` was auto populated.

> Since env vars may contain sensitive information, I wouldn't want to commit them to a file.

Even with the current approach the env vars would be committed to the action workflow yml? My thoughts with the env-vars-file was that if you have alot of vars to populate, it maybe cleaner to just use a file instead of passing them comma separated in the workflow yml.

For secrets, I think the approach should be using secretmanager and retrieving it at runtime directly from the API.

**Comment by arkum** [comment]
Hi @averikitsch as of now firebase CLI doesn't support deploying python cloud functions. Neither the firebase deploy action(s). Wondering whether the best approach will be to make the firebase deploy action include support for python functions as well, which sets the FIREBASE_CONFIG variable automatically

**Comment by arkum** [comment]
Hi @averikitsch as of now firebase CLI doesn't support deploying python cloud functions. Neither the firebase deploy action(s). Wondering whether the best approach will be to make the firebase deploy action include support for python functions as well, which sets the FIREBASE_CONFIG variable automatically

**Comment by averikitsch** [comment]
I see, you are using the python Cloud Functions for now. I'm still hesitant about an env var file and would lean towards improving the env var parsing so you can pass in `env_vars: FIREBASE_CONFIG=${{ secrets.myConfig }}, ...`. It seems like an extra step to have to add the env vars to a file and then pass the file into the action. 

For @arkum , I would deploy the function the first time manually and set the FIREBASE_CONFIG env var. I don't assume this changing on redeployments. Then you'd use the action to do the redeployments which should maintain the env var. Is this correct @bharathkkb ?

**Comment by bharathkkb** [comment]
fixed via #34

**Comment by bharathkkb** [comment]
Best practices for secrets still apply even with the new feature: https://cloud.google.com/functions/docs/env-var#managing_secrets
